File: 80-RemoveDuplicatesFromSortedArray2.py
class Solution:
    def removeDuplicates(self, nums: List[int]):
        x,y,z=0,1,2
        while(z<len(nums)):
            if nums[x] == nums[y] and nums[y] == nums[z]:
                nums.remove(nums[x])
            else:
                x += 1
                y += 1
                z += 1
                
                
# T.C => O(n)
# S.C => O(1)
#approach => Here three pointers are inatialized at the index 0,1,2.
#  Since array is sorted and if all the three elements are same just remove one element from starting index. 
# Continue this process until the last index appears till the last of the array

File: 80-remove_duplicates_from_sortedArray-2.java
//Approach-1: Use LinkedHashMap to store the occurances of the element in Java and then write back
//Approach-2: Remove element and reorder based on the condition of atmost 2 elements
//Approach-3: Using two pointers - fast and slow pointers
//Overall Time Complexity: O(N)
//Overall Space Complexity: O(1)
class Solution {
    public int removeDuplicates(int[] nums) {
        if(nums == null || nums.length == 0) return 0;
        
        int slow = 1;
        int count = 1; //As we are already considering the first element
        
        //Starting from the firt index as we always check the previous element
        for(int fast = 1; fast < nums.length; fast++){
            if(nums[fast] == nums[fast - 1]){
                count++;
            }else{
                count = 1;
            }
            
            if(count <= 2){
                nums[slow] = nums[fast];
                slow++;
            }
        }
        
        return slow;
    }
}